lib/tree.rb

```diff
@@ -16,42 +16,131 @@ def initialize
+    end
+
+    add_helper_method(@root, new_node)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper_method(current, new_node)
+    if new_node.key <= current.key
+      if current.left.nil?
+        return current.left = new_node
+      else
+        add_helper_method(current.left, new_node)
+      end
+    else
+      if current.right.nil?
+        return current.right = new_node
+      else
+        add_helper_method(current.right, new_node)
+      end
+    end 
+  end
+
+
+  # Time Complexity && Space Complexity: O(h) where h is the height of the tree
+  # Tree height has a minimum of O(log n) levels. Worst case, an unbalanced tree height is O(n) levels (https://www.cs.cmu.edu/~adamchik/15-121/lectures/Trees/trees.html)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    current = @root
+
+    while current != nil
+      if key == current.key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+
+    return nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(h) where h is the height of the tree
   def inorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    
+    results = []
+    inorder_helper(@root, results)
+    return results
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, array)
+    return if node.nil?
+    inorder_helper(node.left, array)
+    array << node_reader(node)
+    inorder_helper(node.right, array)
+  end
+
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(h) where h is the height of the tree
+  # reference: https://en.wikipedia.org/wiki/Tree_traversal
   def preorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    
+    results = []
+    preorder_helper(@root, results)
+    return results
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, array)
+    return if node.nil?
+    array << node_reader(node)
+    preorder_helper(node.left, array)
+    preorder_helper(node.right, array)
+  end
+
+  def node_reader(node)
+    return {
+      :key => node.key, 
+      :value => node.value
+    }
+  end
+
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(h) where h is the height of the tree 
   def postorder
-    raise NotImplementedError
+    return [] if @root.nil?
+    
+    results = []
+    postorder_helper(@root, results)
+    return results
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, array)
+    return if node.nil?
+    postorder_helper(node.left, array)
+    postorder_helper(node.right, array)
+    array << node_reader(node)
+  end
+
+  # Time Complexity: O(n) where n is the number of nodes
+  # Space Complexity: O(n) where n is the number of nodes
   def height
```

Comment:
👍 , although the space complexity is O(h) since the recursion maxes out at the height of the tree.